### PR TITLE
Bump snapshot-controller version.

### PR DESCRIPTION
**Pull request by masteryyh**
**Problem:**
Bump `k8s.gcr.io/sig-storage/snapshot-controller` version.

**Solution:**

Edit both `deploy/charts/harvester/values.yaml` and `deploy/charts/harvester/dependency_charts/csi-snapshotter/values.yaml` version.

**Related issue:**
#2983

### CODE REVIEW

File: deploy/charts/harvester/dependency_charts/csi-snapshotter/values.yaml [hunk 1]
```diff
@@ -3,5 +3,5 @@
 # Declare variables to be passed into your templates.
 image:
   repository: k8s.gcr.io/sig-storage/snapshot-controller
-  tag: v4.0.0
+  tag: v6.1.0
```
Comment (guangbochen) [inline deploy/charts/harvester/dependency_charts/csi-snapshotter/values.yaml]: u need to run `helm package update` to update the dependency chart and add the generated tgz file u have modified only.

**Review by guangbochen — APPROVED** [review]
LGTM, thanks.

**Comment by FrankYang0529** [comment]
Looks like there is some permission issue. I can't create VM Snapshot with this PR.

```
> kubectl describe volumesnapshot sn1-volume-vm-disk-0-3kznt
...
  Normal   CreatingSnapshot        110s                      snapshot-controller  Waiting for a snapshot default/sn1-volume-vm-disk-0-3kznt to be created by the CSI driver.
  Warning  SnapshotFinalizerError  <invalid> (x8 over 110s)  snapshot-controller  Failed to check and update snapshot: snapshot controller failed to update default/sn1-volume-vm-disk-0-3kznt on API server: volumesnapshots.snapshot.storage.k8s.io "sn1-volume-vm-disk-0-3kznt" is forbidden: User "system:serviceaccount:kube-system:snapshot-controller" cannot patch resource "volumesnapshots" in API group "snapshot.storage.k8s.io" in the namespace "default"
```

**Comment by FrankYang0529** [comment]
I update clusterrole `snapshot-controller-runner` as following and it works.

```
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  annotations:
    meta.helm.sh/release-name: harvester
    meta.helm.sh/release-namespace: harvester-system
    objectset.rio.cattle.io/id: default-mcc-harvester-cattle-fleet-local-system
  creationTimestamp: "2022-10-19T09:40:15Z"
  labels:
    app.kubernetes.io/managed-by: Helm
    objectset.rio.cattle.io/hash: e852fa897f5eae59a44b4bfe186aad80b10b94b3
  name: snapshot-controller-runner
  resourceVersion: "53441"
  uid: 6f624f1d-4ada-4b04-8137-6dfd498fbf75
rules:
- apiGroups:
  - ""
  resources:
  - persistentvolumes
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - ""
  resources:
  - persistentvolumeclaims
  verbs:
  - get
  - list
  - watch
  - update
- apiGroups:
  - storage.k8s.io
  resources:
  - storageclasses
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - ""
  resources:
  - events
  verbs:
  - list
  - watch
  - create
  - update
  - patch
- apiGroups:
  - snapshot.storage.k8s.io
  resources:
  - volumesnapshotclasses
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - snapshot.storage.k8s.io
  resources:
  - volumesnapshotcontents
  verbs:
  - create
  - get
  - list
  - watch
  - update
  - delete
- apiGroups:
  - snapshot.storage.k8s.io
  resources:
  - volumesnapshots
  verbs:
  - get
  - list
  - watch
  - update
- apiGroups:
  - snapshot.storage.k8s.io
  resources:
  - volumesnapshots
  - volumesnapshots/status
  verbs:
  - update
  - patch
```

**Comment by guangbochen** [comment]
Thanks for doing the validation, chart manifest can be referred to https://github.com/kubernetes-csi/external-snapshotter/blob/master/deploy/kubernetes/snapshot-controller/rbac-snapshot-controller.yaml

**Review by FrankYang0529 — APPROVED** [review]
LGTM

**Comment by bk201** [comment]
Comments from @innobead:

> For current released Longhorn versions, we only support v1beta snapshot-related resources, so would suggest to install compatible versions for CRD and controller, so 6.1 will not be suggested even it still has v1beta1 installed but not served, you can check https://github.com/longhorn/longhorn-manager/pull/1518#discussion_r991818681.
> Some discussion for reference, you can check , https://github.com/longhorn/longhorn/issues/4646#issuecomment-1278260763.
>
>Basically the function will work for sure when using 6.1.0 because v1beta1 is there still, it will be a concern later. This needs more careful to avoid future migration efforts.

**Comment by innobead** [comment]
cc @PhanLe1010 (longhorn upstream)

**Comment by guangbochen** [comment]
As discussed offline, let's bump the CSI `snapshot-controller` to v5.0.1 instead, thanks.

**Comment by guangbochen** [comment]
The CRDs are also updated per v5.0.1 upstream release, PTAL of https://github.com/kubernetes-csi/external-snapshotter/tree/v5.0.1/client/config/crd

**Comment by PhanLe1010** [comment]
> bump the CSI snapshot-controller to v5.0.1 

LGTM

> The CRDs are also updated per v5.0.1 upstream release, PTAL of https://github.com/kubernetes-csi/external-snapshotter/tree/v5.0.1/client/config/crd

Where do we do this? I couldn't see it in this PR

We also need to update our existing code (backup controller/snapshot controller) so that it `creates/reads/updates/deletes` snapshot `v1` instead of `v1beta1`. This is preparing for the removal of `v1beta1` from the serving versions in upstream `v6.1.0` snapshot CRD (future upgrade). Ref:
* https://github.com/harvester/harvester/blob/606acc5249dd012a9b7ead8b32cb2f26de8f246b/pkg/controller/master/backup/backup_status.go#L8
* https://github.com/harvester/harvester/blob/d8247a0341bdc6bc3e0b68ff67fe8200ca4ace1b/pkg/controller/master/backup/util.go#L8
* https://github.com/harvester/harvester/blob/779c158523f9c1dbec3a100f5dadffd86cca9d25/pkg/api/volume/handler.go#L13

**Comment by guangbochen** [comment]
@PhanLe1010 I have verified that the snapshot `v1beta1` still works for the current version(it seems deprecated and not working starting from csi-snapshotter v6.x), we would prefer to bump to `v1` in the next Harvester patch release due to the timeline concerns. cc @FrankYang0529 @bk201 

CRDs are not updated yet and @masteryyh can u help to change it, thanks.

**Comment by masteryyh** [comment]
@PhanLe1010 
I'm doing a test in a 3 nodes rc3 environment, created a volume with snapshot and this volume stuck in `NotReady` state. Use `kubectl describe pvc just-installed`(`just-installed` is the name of volume) can see these warning messages:
```
Warning  ProvisioningFailed    6m6s (x2 over 6m7s)   driver.longhorn.io_csi-provisioner-77b757f445-h5pdn_48bd5495-863a-4a00-b11e-6e950b13d67d  failed to provision volume with StorageClass "longhorn-image-s4h8p": rpc error: code = Internal desc = Bad response statusCode [500]. Status [500 Internal Server Error]. Body: [message=unable to create volume: unable to create volume pvc-193c91fe-88df-42b9-b920-ffb7d681064a: failed to verify data source: cannot get client for volume pvc-55c5ea2b-af1c-4289-b050-ffbac42becf3: engine is not running, code=Server Error, detail=] from [http://longhorn-backend:9500/v1/volumes]
  Warning  ProvisioningFailed    5m38s (x2 over 6m7s)  driver.longhorn.io_csi-provisioner-77b757f445-h5pdn_48bd5495-863a-4a00-b11e-6e950b13d67d  failed to provision volume with StorageClass "longhorn-image-s4h8p": rpc error: code = Internal desc = Bad response statusCode [500]. Status [500 Internal Server Error]. Body: [detail=, message=unable to create volume: unable to create volume pvc-193c91fe-88df-42b9-b920-ffb7d681064a: failed to verify data source: cannot get client for volume pvc-55c5ea2b-af1c-4289-b050-ffbac42becf3: engine is not running, code=Server Error] from [http://longhorn-backend:9500/v1/volumes]
  Normal   Provisioning          114s (x10 over 6m7s)  driver.longhorn.io_csi-provisioner-77b757f445-h5pdn_48bd5495-863a-4a00-b11e-6e950b13d67d  External provisioner is provisioning volume for claim "default/apt-updated"
  Warning  ProvisioningFailed    113s (x6 over 6m2s)   driver.longhorn.io_csi-provisioner-77b757f445-h5pdn_48bd5495-863a-4a00-b11e-6e950b13d67d  failed to provision volume with StorageClass "longhorn-image-s4h8p": rpc error: code = Internal desc = Bad response statusCode [500]. Status [500 Internal Server Error]. Body: [code=Server Error, detail=, message=unable to create volume: unable to create volume pvc-193c91fe-88df-42b9-b920-ffb7d681064a: failed to verify data source: cannot get client for volume pvc-55c5ea2b-af1c-4289-b050-ffbac42becf3: engine is not running] from [http://longhorn-backend:9500/v1/volumes]
  Normal   ExternalProvisioning  11s (x26 over 6m7s)   persistentvolume-controller                                                               waiting for a volume to be created, either by external provisioner "driver.longhorn.io" or manually created by system administrator
```
But after multiple tests I can only reproduce one or two times, most of volumes can be successfully provisioned by longhorn. Is this a known issue for longhorn or maybe a compatibility problem with new `snapshot-controller`?

P.S. maybe easier to reproduce this: attach the volume restored by snapshot to the VM immediately after it's created and start VM. It's still hard to reproduce though.

cc @guangbochen

**Comment by masteryyh** [comment]
UPDATE:
After a lot of tests, it shows that only when attach the volume restored by snapshot while its state is still in NotReady state could this problem appears. Should we wait the volume to be Ready then attach this volume or there is an actual compatibility issue?

Reproduce step:
1. Install a harvester 1.1.0-rc3 environment, with `snapshot-controller` image replaced with `k8s.gcr.io/sig-storage/snapshot-controller:v5.0.1` and CRDs edited according to [here](https://github.com/kubernetes-csi/external-snapshotter/tree/v5.0.1/client/config/crd);
2. Create a VM, with a 10Gi or other size of volume;
4. Create a snapshot of the volume;
5. Use the snapshot to create a volume;
6. While the volume not provisioned by longhorn yet, replace the volume use the restored volume in the VM;
7. Try to boot up the VM, the VM should stuck in `Starting` step;
8. SSH into one of the node, `kubectl describe pvc <volume-name>` should see the warning message given out by longhorn

**Comment by guangbochen** [comment]
@masteryyh we may also need to update our existing code (backup controller/snapshot controller) to the `v1` version instead of `v1beta1` to verify it, if the issue still exists it may sound like an LH bug to me, thanks.

**Comment by PhanLe1010** [comment]
@masteryyh  May I have some questions?

1. Which Longhorn version is it? 

Longhorn only upgrade CSI snapshot stuff in the future version (Longhorn v1.4.0 or Longhorn master-head). So if you are using Longhorn 1.3.2, you shouldn't upgrade CSI snapshot stuff

2. Can you reproduce this issue with CSI snapshot v4.0?

**Comment by masteryyh** [comment]
@PhanLe1010 sorry for delaying for such long time :(

1. Longhorn 1.3.2
2. Still able to reproduce using snapshot-controller:v4.0.0 with longhorn 1.3.2, using steps [here](https://github.com/harvester/harvester/pull/2980#issuecomment-1286486710)

**Comment by PhanLe1010** [comment]
Thanks @masteryyh 
Please create a Longhorn GitHub issue

**Comment by masteryyh** [comment]
Reopening this PR to solve issue https://github.com/harvester/harvester/issues/2983, and for [this](https://github.com/harvester/harvester/pull/2980#issuecomment-1286427486) problem I will open another UI issue to mitigate.

**Review by guangbochen — APPROVED** [review]
LGTM, thanks.

File: go.mod [hunk 2]
```diff
@@ -70,6 +70,7 @@ require (
 	github.com/iancoleman/strcase v0.2.0
 	github.com/k3s-io/helm-controller v0.11.7
 	github.com/k8snetworkplumbingwg/network-attachment-definition-client v0.0.0-20200331171230-d50e42f2b669
+	github.com/kubernetes-csi/external-snapshotter/client/v4 v4.2.0
 	github.com/kubernetes-csi/external-snapshotter/v2 v2.1.3
```
Comment (guangbochen) [inline go.mod]: are we able to remove the v2 library?

**Review by guangbochen — APPROVED** [review]
LGTM, thanks.

**Review by futuretea — DISMISSED** [review]
LGTM, Thanks

**Review by futuretea — COMMENTED** [review]
~- Create a VM in Harvester v1.1.2~
~- Take VM snapshot~
~- Upgrade Harvester cluster to the PR version (build a custom iso from the PR)~
~- Restore the VM snapshot to Create a new VM~
~- The restore VM will stuck in Pending~

~kubectl desribe output of  the pending pvc of the restored new VM:~

**Review by futuretea — APPROVED** [review]
LGTM, The VM Restore issue is an issue in master branch, not introduced by this PR, I will create another issue  https://github.com/harvester/harvester/issues/3943 to track it.

**Review by bk201 — APPROVED** [review]
LGTM, this also fix https://github.com/harvester/harvester-installer/pull/491#issuecomment-1552544638

**Pull request merged**